tensorflow/compiler/mlir/tosa/transforms/legalize_tfl.cc

```diff
@@ -0,0 +1,2918 @@
+      TOSA_REPLACE_LOWERED_OP(rewriter, op, fused_activation_op);
+    }
+  }
+
+  rewriter.replaceOp(op, {conv2d_output});
+
+  return success();
+}
+
+LogicalResult ConvertTFLDepthwiseConv2DOp::matchAndRewrite(
+    Operation* op, PatternRewriter& rewriter) const {
+  auto tfl_conv2d_op = cast<TFL::DepthwiseConv2DOp>(op);
+
+  auto input_type =
+      tfl_conv2d_op.input().getType().dyn_cast<RankedTensorType>();
+  auto filter_type =
+      tfl_conv2d_op.filter().getType().dyn_cast<RankedTensorType>();
+  auto output_type =
+      tfl_conv2d_op.getResult().getType().dyn_cast<RankedTensorType>();
+  // Not a ranked tensor output
+  if (!input_type) return failure();
+  if (!output_type) return failure();
+  if (!filter_type) return failure();
+
+  bool input_is_qtype =
+      input_type.getElementType().isa<mlir::quant::QuantizedType>();
+  bool filter_is_qtype =
+      filter_type.getElementType().isa<mlir::quant::QuantizedType>();
+  bool output_is_qtype =
+      output_type.getElementType().isa<mlir::quant::QuantizedType>();
+
+  if ((input_is_qtype != filter_is_qtype) ||
+      (input_is_qtype != output_is_qtype)) {
+    op->emitOpError(
+        "ConvertTFLConv2DOp: input/filter/output tensor should "
+        "be all quantized or all native");
+    return failure();
+  }
+
+  auto filter_shape = filter_type.getShape();
+  // Operator depthwiseConv2D
+  // TFLite orders the depthwiseConv2D filter in IHWO, while TOSA orders
+  // filter in HWIO
+  //
+  // The lowering reorders the filter.
+  //
+  // a1_transpose = tosa.transpose(filter, {1, 2, 3, 0})   // HWIO
+  // a2_reshape = tosa.reshape(filter, H, W, depth_multiplier, I /
+  // depth_multiplier)
+  // a3_transpose_conv2d = tosa.transpose_conv2d(input, a2_reshape, padding,
+  // stride, dilation)
+
+  auto tmpAttr = tfl_conv2d_op.getAttrOfType<StringAttr>("data_format");
+  if (tmpAttr && tmpAttr.getValue().str() != "NHWC") return failure();
+
+  ArrayAttr pad;
+  ArrayAttr stride;
+  ArrayAttr dilation;
+  StringAttr activation_function;
+
+  auto depth_multiplier =
+      tfl_conv2d_op.getAttrOfType<IntegerAttr>("depth_multiplier");
+
+  {
+    int64_t stride_h = tfl_conv2d_op.stride_h();
+    int64_t stride_w = tfl_conv2d_op.stride_w();
+    stride = rewriter.getI64ArrayAttr({stride_h, stride_w});
+  }
+  {
+    int64_t dilation_h = tfl_conv2d_op.dilation_h_factor();
+    int64_t dilation_w = tfl_conv2d_op.dilation_w_factor();
+    dilation = rewriter.getI64ArrayAttr({dilation_h, dilation_w});
+  }
+  {
+    tensorflow::Padding tf_pad;
+    if (!GetPaddingFromString(tfl_conv2d_op.padding().str(), &tf_pad).ok())
+      return failure();
+
+    if (!getPaddingValuesFromPadType(
+            tf_pad,
+            tensorflow::FORMAT_NHWC,  // TFLite only supports this
+            1,                        // tensorflow::FORMAT_OHWI,
+            input_type, filter_type, stride, dilation, rewriter, pad))
+      return failure();
+  }
+
+  std::vector<int64_t> a1_transpose_dims;
```

Comment:
Here and elsewhere that you are using std::vector for dims: change to `llvm::SmallVector<int64_t, 4>`. For this kind of thing, you can also use `auto a1_transpose_dimes = llvm::to_vector<4>(a, b, c, d)`. There are a bunch of these it would be good to cleanup (not going to comment more).

Reply:
Replaced with SmallVector